File: python-sdk/examples/04_create_swap_order.py
#!/usr/bin/env python3
"""
Example 04: Create Swap Order

Test-backed swap-order flow:
1. Find the RGB asset ID for USDT
2. Get a fresh quote for USDT(RGB_LN) -> BTC(BTC_L1)
3. Ask the local node for a valid BTC receiver address
4. Create a swap order
"""

import asyncio
import logging
import os

from kaleido_sdk import (
    CreateSwapOrderRequest,
    KaleidoClient,
    Layer,
    PairQuoteRequest,
    ReceiverAddress,
    ReceiverAddressFormat,
    SwapLegInput,
)

API_URL = os.getenv("KALEIDO_API_URL", "https://api.staging.kaleidoswap.com")
NODE_URL = os.getenv("KALEIDO_NODE_URL", "http://localhost:3001")

# ---------------------------------------------------------------------------
# Logging setup (application's responsibility — the SDK never does this)
# ---------------------------------------------------------------------------
logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s [%(levelname)-8s] %(name)s — %(message)s",
    datefmt="%H:%M:%S",
)


async def main() -> None:
    """Main entry point."""
    client = KaleidoClient.create(
        base_url=API_URL,
        log_level=logging.DEBUG,
    )
    node_client = KaleidoClient.create(
        base_url=API_URL,
        node_url=NODE_URL,
        log_level=logging.DEBUG,
    )

    print(f"Maker API: {API_URL}")
    print(f"Node URL: {NODE_URL}")
    print("-" * 40)

    print("\nFetching assets...")
    assets = await client.maker.list_assets()
    usdt_asset = next((asset for asset in assets.assets if asset.ticker == "USDT"), None)
    if usdt_asset is None:
        print("USDT asset not found in maker asset list")
        return

    protocol_ids = usdt_asset.protocol_ids or {}
    rgb_asset_id = protocol_ids.get("RGB")
    if rgb_asset_id is None:
        print("USDT RGB asset id not found in protocol_ids")
        return

    print(f"Using RGB asset id for USDT: {rgb_asset_id}")

    print("\nRequesting quote for USDT(RGB_LN) -> BTC(BTC_L1)...")
    quote = await client.maker.get_quote(
        PairQuoteRequest(
            from_asset=SwapLegInput(
                asset_id=rgb_asset_id,
                layer=Layer.RGB_LN,
                amount=1_000_000,
            ),
            to_asset=SwapLegInput(
                asset_id="BTC",
                layer=Layer.BTC_L1,
            ),
        )
    )

    print(f"RFQ ID: {quote.rfq_id}")
    print(f"From: {quote.from_asset.amount} {quote.from_asset.ticker}")
    print(f"To: {quote.to_asset.amount} {quote.to_asset.ticker}")

    print("\nFetching a valid BTC address from the local node...")
    address_response = await node_client.rln.get_address()
    receiver_address = address_response.address
    print(f"Receiver address: {receiver_address}")

    order_request = CreateSwapOrderRequest(
        rfq_id=quote.rfq_id,
        from_asset=quote.from_asset,
        to_asset=quote.to_asset,
        receiver_address=ReceiverAddress(
            address=receiver_address,
            format=ReceiverAddressFormat.BTC_ADDRESS,
        ),
        min_onchain_conf=1,
        refund_address=receiver_address,
        email="test@example.com",
    )

    print("\nCreating swap order...")
    order = await client.maker.create_swap_order(order_request)

    print("\nOrder created:")
    print(f"  Order ID: {order.id}")
    print(f"  RFQ ID: {order.rfq_id}")
    print(f"  Status: {order.status}")
    if order.deposit_address is not None:
        print(f"  Deposit address: {order.deposit_address.address}")
        print(f"  Deposit format: {order.deposit_address.format}")


if __name__ == "__main__":
    asyncio.run(main())

File: typescript-sdk/examples/04_create_swap_order.ts
/**
 * Example 04: Create Swap Order
 *
 * Demonstrates how to create a swap order using the Kaleidoswap SDK.
 * This example shows the complete flow from getting a quote to creating an order.
 */

import {
    KaleidoClient,
    createAssetPairMapper,
    createPrecisionHandler,
    toDisplayAmount,
    Layer,
    ReceiverAddressFormat,
} from 'kaleidoswap-sdk';

const API_URL = process.env.KALEIDO_API_URL || 'http://localhost:8000';

async function main() {
    console.log('🎨 Kaleidoswap SDK - Create Swap Order Example\n');

    const client = KaleidoClient.create({ baseUrl: API_URL });

    // Step 1: Get trading pairs and setup helpers
    console.log('📋 Fetching trading pairs...');
    const pairsResponse = await client.maker.listPairs();

    if (pairsResponse.pairs.length === 0) {
        console.log('No trading pairs available');
        return;
    }

    // Create helpers for asset lookup and precision handling
    const assetMapper = createAssetPairMapper(pairsResponse);
    const precisionHandler = createPrecisionHandler(assetMapper.getAllAssets());

    console.log(`Found ${pairsResponse.pairs.length} trading pairs`);
    console.log(`Mapped ${assetMapper.getAllAssets().length} unique assets\n`);

    // Step 2: Find assets to trade
    const btc = assetMapper.findByTicker('BTC');
    const usdt = assetMapper.findByTicker('USDT');

    if (!btc || !usdt) {
        console.log('❌ BTC or USDT not found in trading pairs');
        console.log('Available assets:', assetMapper.getAllAssets().map((a) => a.ticker).join(', '));
        return;
    }

    console.log(`Found assets:`);
    console.log(`  BTC: ${btc.asset_id}`);
    console.log(`  USDT: ${usdt.asset_id}\n`);

    // Check if they can be traded
    if (!assetMapper.canTrade(btc.asset_id, usdt.asset_id)) {
        console.log('❌ BTC and USDT cannot be traded directly');
        return;
    }

    // Step 3: Find the trading pair and route
    const pair = assetMapper.findPairByTickers('BTC', 'USDT');
    if (!pair || !pair.routes || pair.routes.length === 0) {
        console.log('❌ No routes available for BTC/USDT');
        return;
    }

    const route = pair.routes[0];
    console.log(`Using route: ${route.from_layer} → ${route.to_layer}\n`);

    // Step 4: Prepare amount and validate
    const amountToSwap = 0.0001; // 0.0001 BTC
    const rawAmount = precisionHandler.toRawAmount(amountToSwap, btc.asset_id);

    const validation = precisionHandler.validateOrderSize(amountToSwap, btc);
    if (!validation.valid) {
        console.log(`❌ Amount validation failed: ${validation.error}`);
        return;
    }

    console.log(`💰 Swap amount: ${amountToSwap} BTC (${rawAmount} satoshis)\n`);

    // Step 5: Get a quote
    console.log('💱 Requesting quote...');

    const quote = await client.maker.getQuote({
        from_asset: {
            asset_id: btc.asset_id,
            layer: route.from_layer as Layer,
            amount: rawAmount,
        },
        to_asset: {
            asset_id: usdt.asset_id,
            layer: route.to_layer as Layer,
        },
    });

    console.log('📊 Quote received:');
    console.log(`  RFQ ID: ${quote.rfq_id}`);
    console.log(`  From: ${quote.from_asset.amount} ${quote.from_asset.ticker}`);
    console.log(`  To: ${quote.to_asset.amount} ${quote.to_asset.ticker}`);
    console.log(`  Price: ${quote.price}`);
    console.log(`  Expires: ${new Date(quote.expires_at).toLocaleString()}\n`);

    // Step 6: Create swap order
    // NOTE: In production, use real addresses from your wallet
    console.log('📝 Creating swap order...');

    // Determine receiver address format based on layer
    const receiverFormat: ReceiverAddressFormat =
        route.to_layer === Layer.BTC_LN
            ? ReceiverAddressFormat.BOLT11
            : ReceiverAddressFormat.RGB_INVOICE;

    try {
        const order = await client.maker.createSwapOrder({
            rfq_id: quote.rfq_id,
            from_asset: {
                asset_id: quote.from_asset.asset_id,
                name: btc.name,
                ticker: quote.from_asset.ticker,
                layer: quote.from_asset.layer,
                amount: quote.from_asset.amount,
                precision: btc.precision,
            },
            to_asset: {
                asset_id: quote.to_asset.asset_id,
                name: usdt.name,
                ticker: quote.to_asset.ticker,
                layer: quote.to_asset.layer,
                amount: quote.to_asset.amount,
                precision: usdt.precision,
            },
            receiver_address: {
                // Example placeholder - in production use real address
                address: 'rgb:placeholder-address-for-demo',
                format: receiverFormat,
            },
            min_onchain_conf: 1,
        });

        console.log('✅ Order created successfully!');
        console.log(`  Order ID: ${order.id}`);
        console.log(`  RFQ ID: ${order.rfq_id}`);
        console.log(`  Status: ${order.status}`);

        if (order.deposit_address) {
            console.log(`\n📬 Deposit Address:`);
            console.log(`  Address: ${order.deposit_address.address}`);
            console.log(`  Format: ${order.deposit_address.format}`);
        }

        console.log('\n⏳ Monitoring order status...');

        // Step 7: Monitor order status
        const finalOrder = await client.maker.waitForSwapCompletion(order.id, {
            accessToken: order.access_token,
            timeout: 60000, // 1 minute timeout for demo
            pollInterval: 2000,
            onStatusUpdate: (status) => {
                console.log(`  Status update: ${status}`);
            },
        });

        console.log(`\n🏁 Final status: ${finalOrder.status}`);
    } catch (error) {
        if (error instanceof Error) {
            console.log(`\n⚠️  Order creation failed: ${error.message}`);
            console.log('This is expected in demo mode - provide valid addresses to complete a real swap.');
        } else {
            throw error;
        }
    }

    console.log('\n✅ Demo complete!');
}

main().catch(console.error);
